emission/tests/exportTests/TestExportPipeline.py

```diff
@@ -0,0 +1,78 @@
+from future import standard_library
+standard_library.install_aliases()
+from builtins import *
+import os
+import unittest
+import json
+import bson.json_util as bju
+import pathlib as pl
+import emission.storage.timeseries.abstract_timeseries as esta
+import gzip
+import emission.tests.common as etc
+import emission.pipeline.export_stage as epe
+import emission.storage.pipeline_queries as espq
+import emission.exportdata.export_data as eeed
+import emission.export.export as eee
+
+class TestExportPipeline(unittest.TestCase):
+    def testExportPipeline(self):
```

Comment:
This fine.

However, this is just `testExportMethod` now.  To fully test the pipeline, you will need to add a new `testExportPipeline`, that calls `emission.pipeline.export_stage` and/or `emission/exportdata/export_data`. For that test, you don't need to actually test the file output (since you are already doing so in `testExportMethod`, but the file is generated correctly.

I did say that I would not insist on the refactoring, so you could check for the file existence by simply:
- ensuring that the DATA_DIR is set to a new temporary directory in the setup
- running the pipeline step
- ensuring that there is a file with the UUID and the start_ts in that directory

Since this is a new temporary directory created just for the test, it will only have the file if it was created by the test. Makes sense?

Reply:
Yes. To restate for clarification: 
A second unit test is needed to test that the entire pipeline process is actually working. We would start by setting the environment variable of DATA_DIR to a temporary directory that will only contain the single file we export. We then run the entire pipeline step and check to make sure the file was actually exported to that temporary directory location, and that the naming is correct in reference to the UUID and start_ts.

Reply:
Reaching out with regard to the new unit test I am writing: 

'''
class TestExportPipelineFull(unittest.TestCase):
    def testExportPipelineFull(self):
        #Setup
        etc.setupRealExample(self, "emission/tests/data/real_examples/shankari_2015-07-22") #is this needed? thinking yes
        etc.runIntakePipeline(self.testUUID)
        os.environ['DATA_DIR'] = 'emission/testexports/'

        #Run the full export pipeine
        eeed.export_data(self.testUUID)

        #Verify the export file is in the specified DATA_DIR
        self.assertTrue(path.isdir('emission/testexports/'))
        directory = os.listdir('emission/testexports/')
        self.assertTrue(len(directory) == 1)
'''

Above is what I am thinking for the basic code, not all the tests lined out yet. However I seem to run into the issue that the environment variable is not being set, and the assert statement that the directory exists comes back false. Am a setting the environment variable wrong in this case? Does it not exist outside the script? Not sure why this is becoming an issue. Any help would be appreciated so I can finish this up today.

Reply:
I don't see anything obviously wrong; please use standard debugging steps - e.g. print out the environment variable right after you try to set it, check manually that there is in fact an `emission/testexports` created ...

Also, wrt temporary files/directory, I would prefer creating them in `/tmp` etc using the standard temp directory creation method. Don't write things into the source directory unless you absolutely have to. And if you have to, make sure to clean up once the test is done so it doesn't stick around over the long term.

Reply:
Sounds good. Can you expand on what you mean by the standard temp directory creation method?

Reply:
https://duckduckgo.com/?t=ffsb&q=python+create+temp+directory&ia=web

Reply:
No. I believe this is where my confusion lies. I pass along the uncreated directory path to the export pipeline. Is it then the case that I need to create this directory (a temp directory) before passing along the path to the export pipeline. 

Additionally, if this is the case, it seems straight forward to create the directory temporarily. What is the case for the actual file? As you know that is generated within the the export pipeline itself. I am assuming that is then not created using the tempfile infrastructure you linked.

Reply:
@jruzekowicz, please read the python doc linked above carefully. There is a difference between passing in a string and using the tempfile module. You need to handle both cases properly.

Reply:
Below is the implementation I am working with: 

```
class TestExportPipelineFull(unittest.TestCase):
    def testExportPipelineFull(self):
        #Setup
        etc.setupRealExample(self, "emission/tests/data/real_examples/shankari_2015-07-22") 
        etc.runIntakePipeline(self.testUUID)

        #Using the tempfile module to create a temporary directory
        with tempfile.TemporaryDirectory() as tmpdirname:
            os.environ['DATA_DIR'] = tmpdirname + "/"
            self.assertEqual(os.environ['DATA_DIR'], tmpdirname + "/")
            print(tmpdirname)
            print(os.environ['DATA_DIR'])
            self.assertTrue(path.isdir(tmpdirname))
            eeed.export_data(self.testUUID)
            directory = os.listdir(tmpdirname)
            print(directory)
            self.assertTrue(len(directory) == 1)
```
Here I choose to use the as tmpdirname format after creating a temporary directory. My interpretation here is that after the script finishes what is inside the call, the directory and all if it's content will be cleaned up.  Is this on the right track? 
Additionally, all of the asserts are passing up until I reach the final call, where it appears that ever after the pipeline runs, the directory is left empty. To verify, I also confirmed that the data was not mistakenly going to the archived folder.

Reply:
Yes, this is on the right track for the use of the `tempdir` module. You need to figure out why the directory is left empty.

Note also that you probably want to write a very similar test for a directory created not using `tempdir` to double check directory creation.

Reply:
I am currently trying to understand the difference between using: 
`with tempfile.TemporaryDirectory() as tmpdirname:` vs
`with tempfile.mkdtemp() as tmpdirname:`
It seems like the second would use an absolute path which is not necessarily what I want here, correct me if I am wrong. 

Additionally, can you expand on what you mean by testing directory creation not using tempdir? I am not sure I completely understand the differences between the tempdir module and regular directory, besides the fact that the directory is erased after the function exits.

Reply:
>  It seems like the second would use an absolute path which is not necessarily what I want here, correct me if I am wrong.

No, the documentation says:

"The user of mkdtemp() is responsible for deleting the temporary directory and its contents when done with it."

> can you expand on what you mean by testing directory creation not using tempdir

"This function securely **creates** a temporary directory ". If you pass in a directory like you had before that is not from the `tempdir` module, will it also be created automatically?

Reply:
No, if we pass in a directory that does not exist to the tempdir module, it will not automatically be created for us. I attempted to run: 
```
with tempfile.TemporaryDirectory(dir='emission/nonexistentdir/') as tmpdirname:
            self.assertTrue(path.isdir(tmpdirname))
```
which failed as the directory nonexistentdir did not already exist.

Reply:
correct. So you need two tests:
- one with a tempfile directory (which automatically creates the directory) to ensure that the file is created
- one with a non tempfile directory (which is not automatically created) to ensure that your code creates the directory if it does not exist

You could also combine them into one test; just make sure to clean up the non`tempfile` directory in that case.

Reply:
Okay, to confirm, the code within emission.export should create the directory we declared for the 'DATA_DIR' if it does not already exist. 
Additionally, I am having trouble exporting files to locations not under the emission directory. The following code works: 
```
class TestExportPipelineFull(unittest.TestCase):
    def testExportPipelineFull(self):
        #Setup
        etc.setupRealExample(self, "emission/tests/data/real_examples/shankari_2015-07-22")
        etc.runIntakePipeline(self.testUUID)

        #Create a temporary directory within the emission folder
        with tempfile.TemporaryDirectory(dir='emission/') as tmpdirname:
            self.assertTrue(path.isdir(tmpdirname))

            #Set the envrionment variable
            os.environ['DATA_DIR'] = tmpdirname + "/"
            self.assertEqual(os.environ['DATA_DIR'], tmpdirname + "/")

            #Run the export pipeline
            eeed.export_data(self.testUUID)
            directory = os.listdir(tmpdirname)

            #Check to see if there is a file in the temp directory
            self.assertTrue(len(directory) == 1)

            #Check to make sure the file is of the correct UUID
            uuid = str(self.testUUID)
            file_name = directory[0]
            self.assertTrue(uuid in file_name)
```

However, I attempted to redirect the exporting to happen under the /tmp directory by putting in `dir = ../../../tmp/` and this fails to export the data to that file. It does create the directory but no files are exported to that location. Similarly, the same issue happens when I leave the dir field blank, and just call `temple.TemporaryDirectory()`. The directory itself is created, but the exporting does not take place. Nothing fails, but the file is nowhere to be found. I am stuck on why this happens.